cf-oscore/src/main/java/org/eclipse/californium/oscore/OSCorePeerIdentityHandler.java

```diff
@@ -0,0 +1,74 @@
+/*******************************************************************************
+ * Copyright (c) 2019 RISE SICS and others.
+ * 
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * and Eclipse Distribution License v1.0 which accompany this distribution.
+ * 
+ * The Eclipse Public License is available at
+ *    http://www.eclipse.org/legal/epl-v10.html
+ * and the Eclipse Distribution License is available at
+ *    http://www.eclipse.org/org/documents/edl-v10.html.
+ * 
+ * Contributors:
+ *    Rikard Höglund (RISE SICS)
+ *    
+ ******************************************************************************/
+package org.eclipse.californium.oscore;
+
+import org.eclipse.californium.core.coap.Message;
+import org.eclipse.californium.elements.AddressEndpointContext;
+import org.eclipse.californium.elements.EndpointContext;
+import org.eclipse.californium.elements.auth.OSCorePeerIdentity;
+
+/**
+ * Class that handles the EndpointContext for OSCORE messages. 
+ *
+ * It provides functionality to set an appropriate OSCORE peer identity
+ * for an EndpointContext.
+ *
+ */
+public class OSCorePeerIdentityHandler {
+
+	/**
+	 * Updates the currently set endpoint contexts adding an OSCORE peer identity.
+	 * Will only update endpoint contexts that are set as type AddressEndpointContext
+	 * and have an empty or OSCORE peer identity.
+	 * 
+	 * If for instance Scandium has set another type of EndpointContext or another
+	 * type of Principal as peer identity it will not be updated.
+	 * 
+	 * @param oscoreCtx the OSCORE context used for this message
+	 * @param message the message to set the endpoint context for
+	 */
+	public static void setPeerIdentity(OSCoreCtx oscoreCtx, Message message) {
+
+		//Create a new OSCORE peer identity
+		OSCorePeerIdentity newPeerIdentity = new OSCorePeerIdentity(oscoreCtx.getUri(),
+				oscoreCtx.getIdContext(), oscoreCtx.getRecipientId(), oscoreCtx.getSenderId());
+
+		//Set updated destination context with new OSCORE peer identity
+		EndpointContext dstContext = message.getDestinationContext();
+		if(dstContext instanceof AddressEndpointContext &&
```

Comment:
That would change a MapBasedEndpointContext into a AddressEndpointContext. 
A design, using a method in the context, may be overriden and therefore can keep the MapBasedEndpointContext or AddressEndpointContext.

Reply:
My feeling is, waiting for the other PR #996 would make a implementation easier.

Reply:
> That would change a MapBasedEndpointContext into a AddressEndpointContext.
> A design, using a method in the context, may be overriden and therefore can keep the MapBasedEndpointContext or AddressEndpointContext.

I check that the type of EndpointContext is AddressEndpointContext, otherwise I do not change it. This is to respect the values Scandium sets for DTLS (I also saw that there are different types of EndpoinContext:s that Scandium uses).

Reply:
> My feeling is, waiting for the other PR #996 would make a implementation easier.

Sure, I can wait for that and update my code when it is merged. If there would be a method setPrincipal or similar in the EndpointContext I can use that and not re-create the context with the same values plus a Principal.

Then I could also more easily change the peer identity for any kind of endpoint context type. But I was also worried that if I change it for a DTLS connection it may break some functionality in Scandium.

Currently I only update the context if it is of type AddressEndpointContext and if the existing peer identity is null or of type OSCorePeerIdentity.

Reply:
> I check that the type of EndpointContext is AddressEndpointContext

Though `MapBasedEndpointContext extends AddressEndpointContext` a MapBasedEndpointContext will pass that "instanceof" test.

Reply:
> Though `MapBasedEndpointContext extends AddressEndpointContext` a MapBasedEndpointContext will pass that "instanceof" test.

I see, I did not notice that. Thanks for pointing it out.

I suppose it could be fixed in code to check that it is exactly an AddressEndpointContext but as you mention probably it is better to wait for PR https://github.com/eclipse/californium/pull/996